### PR TITLE
Change Flash retry logic to use backoff delays

### LINKED ISSUE #1890: Flash: Replace dumb retry logic with a truncated exponential backoff algorithm
Most importantly, we need to use randomness (jitter). And we need to add an additional max backoff config option. There's good resources on this here:

https://en.wikipedia.org/wiki/Exponential_backoff
https://cloud.google.com/iot/docs/how-tos/exponential-backoff
https://docs.microsoft.com/en-us/azure/architecture/best-practices/transient-faults
https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

A part of this was implemented by @omerfirmak (thanks!), but it's not used consistently across the code and it lacks the truncation & jitter.

Edit: This approach is needed not just for connection errors (which would ideally be handled in vibe.d), but also race conditions between two flash nodes. In particular the nodes may not have the exact same sequence ID / channel state / block height at any given time - therefore the nodes should attempt to retry their requests when it makes sense. If for example the sequence ID doesn't match (yet), the request should be retried. But if the node sends an invalid signature, then the request doesn't need to be retried because it shows there was a logical error or the node was byzantine.

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This also includes a bunch of fixes to the internals.

But let's see what the CI says.

Fixes #1890

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
> submodules/agora/source/agora/flash/UpdateSigner.d(35,8): Error: module `Backoff` is in file 'agora/utils/Backoff.d' which cannot be read

So every time we add a new module we first have to update Faucet's `dub.json` file and add that file there, even without knowing whether the PR will be accepted. This is ridiculous.

File: source/agora/flash/Channel.d [hunk 1]
```diff
@@ -1157,13 +1185,24 @@ LOuter: while (1)
     {
         const old_balance = this.cur_balance;
 
+        const old_seq_id = this.cur_seq_id;
```
Comment (omerfirmak) [inline source/agora/flash/Channel.d]: `collectSignatures` already takes `update.seq_id` as an argument. Could we wait until `collectSignatures` to return to update the `cur_seq_id`?

File: source/agora/flash/Node.d [hunk 2]
```diff
@@ -853,10 +853,11 @@ public abstract class FlashNode : ControlFlashAPI
         // our own secret (we are the payee)
         if (auto secret = payment_hash in this.secrets)
         {
-            assert(!error); // Payee should never fail to receive
-            channel.learnSecrets([*secret], [], this.last_block_height);
+            if (!error)
```
Comment (omerfirmak) [inline source/agora/flash/Node.d]: I can't see how, is there an example you can give? AFAICS, `onPaymentComplete` is never called with an error for incoming payments anyways.

File: source/agora/flash/Channel.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/flash/Channel.d]: Sure, good idea.

File: source/agora/flash/Node.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/flash/Node.d]: The signing process can fail. I added a notification for failed incoming payments, but now that I think about it maybe it should be a separate callback.

I'll rework it a bit.

**Comment by AndrejMitrovic** [comment]
Thanks for the tips @omerfirmak. I removed the incoming payment change and fixed up the sequences.

**Comment by AndrejMitrovic** [comment]
It looks like https://github.com/bosagora/faucet/pull/56 isn't ready yet so we'll have to do this in a funky way by juggling submodules..

**Comment by AndrejMitrovic** [comment]
Edit: s/submodules/docker images

**Comment by AndrejMitrovic** [comment]
It might be easier to just merge it first, then fix faucet to add the new import, and then update agora again. Let's wait for green from the CI.

**Comment by AndrejMitrovic** [comment]
Merging this, and I'll fix up the faucet version soon.

**Pull request merged**